### PR TITLE
test(agent): remove race in TestAgent_Session_TTY_QuietLogin/Hushlogin

### PR DESCRIPTION
**Pull request by zedkipp**
The subtest previously called session.Shell(), wrote "exit 0" through a client-side PTY, and then waited indefinitely on session.Wait(). Under the race detector the byte stream occasionally arrived at the agent before the remote shell was in its read loop and was silently discarded; the shell never exited, session.Wait() blocked until the go-test watchdog kicked in and killed the test binary.

The agent writes the message of the day announcement banner synchronously in agentssh.startPTYSession before forking the user shell. The subtest now observes the banner on the SSH session's output stream via testutil.WaitBuffer and asserts that the message of the day never appears in the same stream. This change removes the input-side race entirely so the test no longer depends on the remote shell reading stdin or exiting cleanly.

Also fixes a pre-existing test bug where the empty bytes intended to create ~/.hushlogin were written to the MOTD path. The previous test passed only because the MOTD file ended up empty, not because the hushlogin code path was exercised. With the file now placed at the correct path, the assertion genuinely validates isQuietLogin.

Related to coder/internal#1509.

Generated with assistance from Coder Agents.

### CODE REVIEW

File: agent/agent_test.go [hunk 1]
```diff
@@ -953,35 +954,53 @@ func TestAgent_Session_TTY_QuietLogin(t *testing.T) {
...
+		// drive the test entirely off those writes, observed on the SSH
+		// session's output stream, and never depend on the remote shell
+		// reading our stdin or exiting cleanly.
+		//
+		// Earlier revisions of this subtest wrote "exit 0" to a client
+		// PTY and then called session.Wait(). Under -race that pattern
+		// occasionally hangs forever: the bytes arrive at the agent
+		// before the shell is in its read loop and get silently dropped,
+		// so the shell never exits and the go test watchdog kills the
+		// binary.
```
Comment (mafredri) [inline agent/agent_test.go]: This comment is too wordy and should be removed. (I'm actually in the process of implementing a redesign of our test suite to avoid the issue where you write to the shell before it's ready.)

File: agent/agent_test.go [hunk 2]
```diff
@@ -953,35 +954,53 @@ func TestAgent_Session_TTY_QuietLogin(t *testing.T) {
...
-		err = session.Wait()
-		require.NoError(t, err)
-
-		require.NotContains(t, stdout.String(), wantNotMOTD, "should not show motd")
-		require.Contains(t, stdout.String(), wantMaybeServiceBanner, "should show service banner")
+		require.NoError(t, session.Shell())
+
+		bannerCtx, bannerCancel := context.WithTimeout(context.Background(), testutil.WaitShort)
+		defer bannerCancel()
+		stdout.RequireWaitFor(bannerCtx, t, wantMaybeServiceBanner)
```
Comment (mafredri) [inline agent/agent_test.go]: While we're here we should probably rename `wantMaybeServiceBanner` to just `wantServiceBanner`, it seems to be used by all subtests.

File: agent/agent_test.go [hunk 3]
```diff
@@ -953,35 +954,53 @@ func TestAgent_Session_TTY_QuietLogin(t *testing.T) {
...
-		ptty.WriteLine("exit 0")
-		err = session.Wait()
-		require.NoError(t, err)
-
-		require.NotContains(t, stdout.String(), wantNotMOTD, "should not show motd")
-		require.Contains(t, stdout.String(), wantMaybeServiceBanner, "should show service banner")
+		require.NoError(t, session.Shell())
+
+		bannerCtx, bannerCancel := context.WithTimeout(context.Background(), testutil.WaitShort)
+		defer bannerCancel()
```
Comment (mafredri) [inline agent/agent_test.go]: Suggestion: Use `testutil.Context` here.

File: agent/agent_test.go [hunk 4]
```diff
@@ -953,35 +954,53 @@ func TestAgent_Session_TTY_QuietLogin(t *testing.T) {
...
+		stdout.RequireWaitFor(bannerCtx, t, wantMaybeServiceBanner)
+
+		// Banner is written immediately before MOTD on the agent (see
+		// agentssh.startPTYSession). If MOTD were going to appear it
+		// would arrive within milliseconds of the banner, so a short
+		// Never window is sufficient to confirm hushlogin took effect.
+		require.Never(t, func() bool {
+			return strings.Contains(stdout.String(), wantNotMOTD)
+		}, testutil.IntervalSlow, testutil.IntervalFast, "should not show motd")
```
Comment (mafredri) [inline agent/agent_test.go]: Wrong values here, one should be duration rather than interval. But I'd love to have a different way to detect this rather than waiting a predetermined amount of time (flakey when doesn't wait long enough, and wasteful because waits longer than necessary).

As mentioned earlier, I'm working on something that will make "exit 0" above work reliably, but that doesn't help you now. I'd propose the following:

```go
stdin, err := session.StdinPipe()
require.NoError(t, err)
stdout := testutil.NewWaitBuffer()
session.Stdout = stdout
require.NoError(t, session.Shell())

testutil.Go(t, func() {
    for {
        if _, err := stdin.Write([]byte("exit 0\n")); err != nil {
            return
        }
		time.Sleep(testutil.IntervalFast)
    }
})

err = session.Wait()
require.NoError(t, err)

require.NotContains(t, stdout.String(), wantNotMOTD, "should not show motd")
require.Contains(t, stdout.String(), wantMaybeServiceBanner, "should show service banner")
```

The sleep is optional I suppose, but I don't think spamming will help.

File: agent/agent_test.go [hunk 5]
```diff
@@ -924,17 +924,18 @@ func TestAgent_Session_TTY_QuietLogin(t *testing.T) {
 	u, err := user.Current()
 	require.NoError(t, err, "get current user")
 
-	name := filepath.Join(u.HomeDir, "motd")
+	motdPath := filepath.Join(u.HomeDir, "motd")
+	hushloginPath := filepath.Join(u.HomeDir, ".hushlogin")
```
Comment (mafredri) [inline agent/agent_test.go:928]: Good rename! The `name` variable confusion is what caused this test to become wrong in the afero refactor.

File: agent/agent_test.go (hunk 1 above)
Comment (zedkipp) [reply agent/agent_test.go]: Removed. Lately I've been leaving more "why" in comments because there's so much test copy-pasta and prior-art influence by agents.

File: agent/agent_test.go (hunk 4 above)
Comment (zedkipp) [reply agent/agent_test.go]: I went with your suggested approach of retrying the "exit 0", but I bounded the for loop and the `session.Wait()`.

In the absence of a mechanism that eliminates the possibility for these early writes, I think what we have is decent.

File: agent/agent_test.go [hunk 6]
```diff
@@ -947,41 +948,65 @@ func TestAgent_Session_TTY_QuietLogin(t *testing.T) {
...
+					return
+				}
+				select {
+				case <-deadlineCh:
+					t.Errorf("failed to exit shell within %s", retryDur)
+					return
+				case <-ticker.C:
+				}
+			}
+		})
```
Comment (mafredri) [inline agent/agent_test.go:991]: I feel this adds a lot of complexity to what is essentially a simple context check. Consider the following:

```go
ctx := testutil.Context(t)
context.AfterFunc(ctx, func() { _ = session.Close() })

testutil.Go(t, func() {
    for {
        if _, err := stdin.Write([]byte("exit 0\n")); err != nil {
            return
        }
    }
})

err = session.Wait()
require.NoError(t, err)
```

Write and Wait will both be interrupted when the session is closed.

Comment (zedkipp) [reply agent/agent_test.go:991]: I always forget about `context.AfterFunc()`. This is so much cleaner, thanks.

File: agent/agent_test.go [hunk 7]
```diff
@@ -947,41 +948,53 @@ func TestAgent_Session_TTY_QuietLogin(t *testing.T) {
...
 		})
 		err = session.RequestPty("xterm", 128, 128, ssh.TerminalModes{})
 		require.NoError(t, err)
 
+		stdout := testutil.NewWaitBuffer()
 		ptty := ptytest.New(t)
-		var stdout bytes.Buffer
-		session.Stdout = &stdout
+		session.Stdout = stdout
 		session.Stderr = ptty.Output()
```
Comment (mafredri) [inline agent/agent_test.go:976]: We only use ptty to capture stderr output here, not harmful, but could potentially be removed. I'm fine either way though.

**Review by mafredri — APPROVED** [review]
:shipit:
